docs/public/specifications/test-zones/Engine/Recursor-PM/CNAME.md

```diff
@@ -0,0 +1,236 @@
+* The query name is given by the scenario.
+* Query type is assumed to be `A`.
+* In the zone, the query name always has a `CNAME` record.
+* The zone is set up on one NS, ns1.
+* The [RCODE Name] in the response is NoError.
+
+### GOOD-CNAME-1
+The query name will resolve to one `A` record via one CNAME.
+
+* Query name: "good-cname-1.cname.recursor.engine.xa"
+  * To be found in the answer section:
+```
+   good-cname-1         CNAME good-cname-1-target 
+   good-cname-1-target  A     127.0.0.1
+```
+
+### GOOD-CNAME-2
+The query name will resolve to two `A` record via one CNAME.
+
+* Query name: "good-cname-2.cname.recursor.engine.xa"
+  * To be found in the answer section:
+```
+   good-cname-2         CNAME good-cname-2-target
+   good-cname-2-target  A     127.0.0.1
+   good-cname-2-target  A     127.0.0.2
+```
+
+### GOOD-CNAME-CHAIN
+The query name will resolve to two `A` record via three CNAME.
+
+* Query name: "good-cname-chain.cname.recursor.engine.xa"
+  * To be found in the answer section:
+```
+   good-cname-chain         CNAME good-cname-chain-two
+   good-cname-chain-two     CNAME good-cname-chain-three
+   good-cname-chain-three   CNAME good-cname-chain-target
+   good-cname-chain-target  A     127.0.0.1
+```
+
+### GOOD-CNAME-OUT-OF-ZONE
+The query name will resolve to an `A` record via a CNAME in the zone with a
+target that points at a subzone, and a delegation to that sub zone.
+
+* For all queries
+  * The two zones, `cname.recursor.engine.xa` and
+    `goodsub.cname.recursor.engine.xa`, are hosted on different IP addresses.
+  * "x", "y" and "z" in the IP addresses in the configuration below are to be
+    set in the configuration of the test zones.
+
+* Query name: "good-cname-out-of-zone.cname.recursor.engine.xa"
+  * Servers: NS of `cname.recursor.engine.xa`.
+  * Answer, authority and additional sections, respectively, to be found in the
+    response:
+```
+   ;; ANSWER SECTION:
+   good-cname-out-of-zone  CNAME target.goodsub
+
+   ;; AUTHORITY SECTION:
+   goodsub                 NS    ns1.goodsub
+
+   ;; ADDITIONAL SECTION:
+   ns1.goodsub             A     127.x.y.z
+   ns1.goodsub             AAAA  fda1:b2:c3::127:x:y:z
+```
+* Query name: "target.goodsub.cname.recursor.engine.xa"
+  * Servers: NS of `goodsub.cname.recursor.engine.xa`.
+  * To be found in the answer section:
+```
+   target.goodsub          A     127.0.0.1
+```
+
+### NXDOMAIN-VIA-CNAME
+The query name exists, but as CNAME record. The target name of CNAME does not
+exist.
+
+* Query name: "nxdomain-via-cname.cname.recursor.engine.xa"
+  * The [RCODE Name] in the response is NxDomain.
+  * The target of the CNAME, `nxdomain-via-cname-target` does not exist in the,
+    zone and is not delegated.
+  * To be found in the answer section:
+```
+   nxdomain-via-cname       CNAME nxdomain-via-cname-target
+```
+
+### NODATA-VIA-CNAME
+The query name exists, but as CNAME record. The target name of CNAME does not
+exist.
```

Comment:
The target name exists as a node, but does not contain any resource records of type A.